bundles/org.openhab.binding.homeconnect/src/main/java/org/openhab/binding/homeconnect/internal/client/OkHttpHelper.java

```diff
@@ -0,0 +1,170 @@
+import static org.openhab.binding.homeconnect.internal.HomeConnectBindingConstants.HTTP_PROXY_HOST;
+import static org.openhab.binding.homeconnect.internal.HomeConnectBindingConstants.HTTP_PROXY_PORT;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.net.Proxy;
+import java.time.Duration;
+import java.time.LocalDateTime;
+
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.SSLSocketFactory;
+import javax.net.ssl.TrustManager;
+import javax.net.ssl.X509TrustManager;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.jetty.http.HttpMethod;
+import org.openhab.binding.homeconnect.internal.client.exception.AuthorizationException;
+import org.openhab.binding.homeconnect.internal.client.exception.CommunicationException;
+import org.openhab.binding.homeconnect.internal.client.exception.ProxySetupException;
+import org.openhab.core.auth.client.oauth2.AccessTokenResponse;
+import org.openhab.core.auth.client.oauth2.OAuthClientService;
+import org.openhab.core.auth.client.oauth2.OAuthException;
+import org.openhab.core.auth.client.oauth2.OAuthResponseException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+
+import io.github.bucket4j.Bucket;
+import io.github.bucket4j.Bucket4j;
+import okhttp3.OkHttpClient;
+import okhttp3.OkHttpClient.Builder;
+import okhttp3.Request;
+
+/**
+ * okHttp helper.
+ *
+ * @author Jonas Brüstel - Initial contribution
+ *
+ */
+@NonNullByDefault
+public class OkHttpHelper {
+    private static final String HEADER_AUTHORIZATION = "Authorization";
+    private static final String BEARER = "Bearer ";
+    private static final int OAUTH_EXPIRE_BUFFER = 10;
+    private static final JsonParser JSON_PARSER = new JsonParser();
+    private static final Gson GSON = new GsonBuilder().setPrettyPrinting().create();
+    private static final Logger LOGGER = LoggerFactory.getLogger(OkHttpHelper.class);
+    private static final Bucket BUCKET = Bucket4j.builder()
+            // allows 50 tokens per minute (added 10 second buffer)
+            .addLimit(classic(50, intervally(50, Duration.ofSeconds(70))).withInitialTokens(40))
+            // but not often then 50 tokens per second
+            .addLimit(classic(10, intervally(10, Duration.ofSeconds(1))).withInitialTokens(0)).build();
+
+    public static Builder builder(boolean enableRateLimiting) {
+        Builder builder;
+        if (HTTP_PROXY_ENABLED) {
+            LOGGER.warn("Using http proxy! {}:{}", HTTP_PROXY_HOST, HTTP_PROXY_PORT);
+            Proxy proxy = new Proxy(Proxy.Type.HTTP, new InetSocketAddress(HTTP_PROXY_HOST, HTTP_PROXY_PORT));
+
+            try {
+                TrustManager[] trustAllCerts = new TrustManager[] { new X509TrustManager() {
```

Comment:
Is this meant to stay this way?
It shouldn't.

Reply:
Removed it. I used it to reproduce error cases while using a proxy.